src/transformers/models/superpoint/modeling_superpoint.py

```diff
@@ -0,0 +1,577 @@
+        if k >= len(keypoints):
+            return keypoints, scores
+        scores, indices = torch.topk(scores, k, dim=0)
+        return keypoints[indices], scores
+
+
+class SuperPointDescriptorDecoder(nn.Module):
+    """
+    The SuperPointDescriptorDecoder uses the outputs of both the SuperPointEncoder and the
+    SuperPointInterestPointDecoder to compute the descriptors at the keypoints locations.
+
+    The descriptors are first computed by a convolutional layer, then normalized to have a norm of 1. The descriptors
+    are then interpolated at the keypoints locations.
+    """
+
+    def __init__(self, config: SuperPointConfig) -> None:
+        super().__init__()
+        self.hidden_sizes = config.hidden_sizes
+        self.descriptor_dim = config.descriptor_dim
+        self.keypoint_threshold = config.keypoint_threshold
+        self.max_keypoints = config.max_keypoints
+        self.nms_radius = config.nms_radius
+        self.border_removal_distance = config.border_removal_distance
+
+        self.relu = nn.ReLU(inplace=True)
+        self.pool = nn.MaxPool2d(kernel_size=2, stride=2)
+        self.convDa = nn.Conv2d(
+            self.hidden_sizes[3],
+            self.hidden_sizes[4],
+            kernel_size=3,
+            stride=1,
+            padding=1,
+        )
+        self.convDb = nn.Conv2d(
+            self.hidden_sizes[4],
+            self.descriptor_dim,
+            kernel_size=1,
+            stride=1,
+            padding=0,
+        )
+
+    def forward(self, encoded, keypoints) -> torch.Tensor:
+        """Compute the dense descriptors"""
+        descriptors = self.convDb(self.relu(self.convDa(encoded)))
+        descriptors = torch.nn.functional.normalize(descriptors, p=2, dim=1)
+
+        # Extract descriptors
+        descriptors = self.sample_descriptors(keypoints[None], descriptors[0][None], 8)[0]
+
+        # From [descriptor_dim, num_keypoints] to [num_keypoints, descriptor_dim]
+        descriptors = torch.transpose(descriptors, 0, 1)
+
+        return descriptors
+
+    @staticmethod
+    def sample_descriptors(keypoints, descriptors, scale: int = 8) -> torch.Tensor:
+        """Interpolate descriptors at keypoint locations"""
+        batch_size, num_channels, height, width = descriptors.shape
+        keypoints = keypoints - scale / 2 + 0.5
+        divisor = torch.tensor([[(width * scale - scale / 2 - 0.5), (height * scale - scale / 2 - 0.5)]])
+        divisor = divisor.to(keypoints)
+        keypoints /= divisor
+        keypoints = keypoints * 2 - 1  # normalize to (-1, 1)
+        kwargs = {"align_corners": True} if torch.__version__ >= "1.3" else {}
+        descriptors = torch.nn.functional.grid_sample(
+            descriptors, keypoints.view(batch_size, 1, -1, 2), mode="bilinear", **kwargs
+        )
+        descriptors = torch.nn.functional.normalize(descriptors.reshape(batch_size, num_channels, -1), p=2, dim=1)
```

Comment:
Lets split this across more lines -it'll make future debugging easier. Could you add comments descripting the shape changes from`reshape`? 

```suggestion
        # FILLME [x, y, z] -> [p,q ,r]
        keypoints = keypoints.view(batch_size, 1, -1, 2)
        descriptors = torch.nn.functional.grid_sample(descriptors, , mode="bilinear", **kwargs)
        # FILLME [x, y, z] -> [p,q ,r]
        descriptors.reshape(batch_size, num_channels, -1)
        descriptors = torch.nn.functional.normalize(descriptors, p=2, dim=1)
```

Reply:
Fixed in [this commit](https://github.com/huggingface/transformers/pull/25786/commits/1bd37a0ed8f038a47cb32a23b911200bb23b43fa)